prefetch.bs

```diff
@@ -0,0 +1,219 @@
+        1. If |request|'s [=request/reserved client=] is null, then:
+            1. Let |topLevelCreationURL| be |currentURL|.
+            1. Let |topLevelOrigin| be null.
+            1. If |browsingContext| is not a [=top-level browsing context=], then:
+                1. Let |parentEnvironment| be |browsingContext|'s [=browsing context/container=]'s [=relevant settings object=].
+                1. Set |topLevelCreationURL| to |parentEnvironment|'s [=environment/top-level creation URL=] and |topLevelOrigin| to |parentEnvironment|'s [=environment/top-level origin=].
+            1. Set |request|'s [=request/reserved client=] to a new [=environment=] whose [=environment/id=] is a unique opaque string, [=environment/target browsing context=] is |browsingContext|, [=environment/creation URL=] is |currentURL|, [=environment/top-level creation URL=] is |topLevelCreationURL|, and [=environment/top-level origin=] is |topLevelOrigin|.
+        1. Let |proposedPartitionKey| be the result of [=determining the network partition key=] given |request|.
+        1. If |partitionKey| is not equal to |proposedPartitionKey|, then set |response| to a [=network error=] and [=iteration/break=].
+
+            <div class="issue">It might be possible to "downgrade" to [=uncredentialed prefetch=] in this case.</div>
+        1. TODO: this is where [=process a navigate fetch=] checks the `navigate-to` CSP
+        1. If |response| is null, [=fetch=] |request|.
+        1. Otherwise, perform [=HTTP-redirect fetch=] using |request| and |response|.
+        1. Wait for the [=task=] on the [=networking task source=] to [=process response=] and set |response| to the result.
+        1. TODO: Figure out if the coop enforcement logic is necessary here; it probably is
+        1. TODO: Figure out if the corp enforcement logic is necessary here; it probably is
+        1. Set |locationURL| to |response|'s [=response/location URL=].
+        1. If |locationURL| is not a [=URL=] whose [=url/scheme=] is an [=HTTP(S) scheme=], then [=iteration/break=].
+    1. If |locationURL| is failure or a [=URL=] whose [=url/scheme=] is not an [=HTTP(S) scheme=], then set |response| to a [=network error=].
+    1. TODO: navigate-to, frame-src, XFO enforcement should probably be left to navigation, but what about status codes and Content-Disposition?
+    1. If |response| is a [=network error=], then return.
+    1. [=Store a prefetch record=] given |document|, |url|, |referrerPolicy| and |response|.
+</div>
+
+The <dfn>list of sufficiently strict speculative navigation referrer policies</dfn> is a list containing the following: "", "`strict-origin-when-cross-origin`", "`strict-origin`", "`same-origin`", "`no-referrer`".
+
+<div algorithm="uncredentialed prefetch">
+    To <dfn export>uncredentialed prefetch</dfn> given a [=document=] |document|, [=URL=] |url| and [=referrer policy=] |referrerPolicy|, perform the following steps.
```

Comment:
Any way to share more logic between this and the previous, so that it's obvious what the differences are?

Reply:
Maybe; I can make another attempt. I'm a little worried that if it becomes too littered with "if you're in mode X, do Y" checks then it will become harder to reason about the properties of each.

Reply:
Yeah, very fair concern. It might be that just getting the redirect stuff factored out is enough.